docs/faas/aws-lambda.md

```diff
@@ -54,21 +54,29 @@ and the [cloud resource conventions][cloud]. The following AWS Lambda-specific a
-resulting `Context` is [valid](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/trace/api.md#isvalid) then a [Span Link][] SHOULD be added to the new Span's
-[start options](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/trace/api.md#specifying-links) with an associated attribute of `source=x-ray-env` to
-indicate the source of the linked span.
-Instrumentation MUST check if the context is valid before using it because the `_X_AMZN_TRACE_ID` environment variable can
-contain an incomplete trace context which indicates X-Ray isn’t enabled. The environment variable will be set and the
-`Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
-disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
+When [AWS X-Ray Active Tracing](https://docs.aws.amazon.com/lambda/latest/dg/services-xray.html) is enabled for a Lambda,
+the runtime will automatically generate a span based on configured sampling rates and propagate the span context
+via the `_X_AMZN_TRACE_ID` environment variable (and the `com.amazonaws.xray.traceHeader` system property for Java Lambda functions).
+This span context is encoded using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/context/api-propagators.md).
 
-**Note**: When instrumenting a Java AWS Lambda, instrumentation SHOULD first try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) before checking and attempting to parse the environment variable above.
+Users MUST be able to configure the propagator to prioritize propagating the X-Ray "Active Tracing" span context.
+(Users probably want this enabled if OpenTelemetry is configured to report spans to AWS X-Ray so their trace is linked together properly.)
 
-[Span Link]: https://opentelemetry.io/docs/concepts/signals/traces/#span-links
+Implementations MUST provide an additional propagator configurable as `xray-lambda` which ignores the given carrier instance and instead attempts to propagate
+the span context from the `_X_AMZN_TRACE_ID` environment variable (and the `com.amazonaws.xray.traceHeader` system property for Java Lambda functions with priority given
+to the system property if set).
+
+Since propagators are invoked in order, users would give priority to X-Ray's "Active Tracing" span context by setting the environment variable:
+
+`OTEL_PROPAGATORS=tracecontext,baggage,xray,xray-lambda`
+
+To avoid broken traces, if OpenTelemetry is reporting traces to another system besides AWS X-Ray, users should either omit `xray-lambda` or add it to the beginning:
```

Comment:
I don't think this approach can work. If xray-lambda is set as global propagator, you are blocked from extract()ing anything else anywhere in the Lambda function (e.g. what if your Lambda polls a messagig queue? All the links/parents will be overriden with what is extraced from _X_AMZN_TRACE_ID.

I think this needs to be an option for the AWS Lambda instrumentation, not a global option. Or otherwise the xray-lambda pseudo-propagator needs to be no-op also for extraction when called for anything else than the extract for the parent of the Lambda handler (which seems more messy to implement and harder to understand than an ordinary instrumentation option).

Please tell me if I misunderstood something.

Reply:
The .NET Lambda instrumentation currently has something similar as option on the instrumentation: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AWSLambda#awslambdainstrumentationoptions

However, the downside of this is that you can't _globally_ disable or enable all context propagation. The [context propagation spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#global-propagators) calls this approach a "discouraged practice":

> Note: It is a discouraged practice, but certain instrumentation libraries might use proprietary context propagation protocols or be hardcoded to use a specific one. In such cases, instrumentation libraries MAY choose not to use the API-provided propagators and instead hardcode the context extraction and injection logic.

Reply:
> However, the downside of this is that you can't globally disable or enable all context propagation. 

To be clear: I have nothing against the global OTEL_PROPAGATORS option itself (this is not being discussed in this PR, as it already exists IIUC). My concern is that the use of this global option to solve a problem local to the AWS Lambda instrumentation will break all other instrumentations (or user code) calling extract in other contexts.

Reply:
How does this break code in other places?

Reply:
Please see my first post. If that is unclear, I can try to provide pseudo-code.

Reply:
Let's say I have a Lambda function that is timer-triggered. Every 10 minutes it polls all messages from some queue (maybe SQS, maybe a custom REST-based API). When the function is triggered, I want to have the X-Ray span as parent of the incoming FaaS span, so I set the global propagator to xray-lambda. So far so good. Now I (or the AWS SDK or custom REST messaging instrumentation) will call GetGlobalPropagator().Extract(message.headers) to extract a link to put on the message. Because the global propagator is configured to xray-lambda, this succeeds for every message. Every message will be linked to the X-Ray span for the event handler. If the messages actually had some other messaging headers that contained a span context, it will be lost.

Reply:
Based on the updated example you've provided above, I don't think the previous status quo (prior to this year's changes) would have handled this well either.

That being said, there are [proposals](https://github.com/open-telemetry/opentelemetry-java/pull/5819) to improving the extract behavior involving conflicting propagators (basically span link everything else).  I think that plus my suggestion above to skip if a span is already active should handle your example sufficiently.

Reply:
> Based on the updated example you've provided above, I don't think the previous status quo (prior to this year's changes) would have handled this well either.

I disagree. The current status quo has other problems, but none so severe.

> That being said, there are https://github.com/open-telemetry/opentelemetry-java/pull/5819 to improving the extract behavior involving conflicting propagators (basically span link everything else). I think that plus my suggestion above to skip if a span is already active should handle your example sufficiently.

I think that will make everybody suffer increased complexity for the sake of the AWS Lambda instrumentation. IMHO this is an AWS-Lambda-specific configuration need that is best handled as configuration options for the AWS Lambda instrumentation.

Specifically, how would you solve this with additional links (pseudo-code)? It will need special casing in the sensor anyways.

```python
def extract_parent_from_event(event):
    if event.type == "ApiGateway":
        return extract_http(event)
    elif event.type == '...':
        return extract_from_...(event)
    else:
        # In case of SQS we would not normally set any parent (messaging spec), so
        # we would lose the parent if active tracing is configured without a special
        # instrumentation option
        return Context.ROOT

def instrumented_handle_request(event, lambda_context):
    spanOptions = SpanOptions(name=lambda_context.function_name, parent=extract_parent_from_event(event))

    if event.type == "SQS":
        for message in event.messages:
            # Here, a batch with N messages will have N links poiting to the
            # X-Ray-created parent if the global propagator is configured to
            # xray-lambda. The get_message_header function would never even be
            # called.
            SpanOptions.links.append(Link(propagators.get_global().extract(message.headers, get_message_header, Context.ROOT)))
    with spanOptions.start_active_span() as span:
        return original_handle_request(event, lambda_context)
```

Reply:
> I disagree. The current status quo has other problems, but none so severe.

I was referring to the previous solution of always prioritizing the env var context, but handled in instrumentation. Not the current spec for using span links.

> I think that will make everybody suffer increased complexity for the sake of the AWS Lambda instrumentation. IMHO this is an AWS-Lambda-specific configuration need that is best handled as configuration options for the AWS Lambda instrumentation.

I see that proposal as a generically beneficial change and not specific to (or required by) lambda.


---------------------

As for your example above, note that it does not conform to our current [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/faas/aws-lambda.md#sqs-message). Specifically those span links should be on a message span, not the event span. That same document also specifies that link propagation should be done specifically using the x-ray propagator rather than the global propagator because x-ray is the only option AWS supports for more complicated interaction patterns.

That said, assuming the following proposal were implemented:

> I would be open to considering the following addition: "the xray-lambda propagator extract MUST return the context unmodified if the context already has an active span". That seems straightforward to implement and should address your concern.

I would then approach this by putting the `parent` into a context that is passed into extract (instead of `Context.ROOT`).  I agree not an ideal solution, but still workable.

---------------

@Oberon00 perhaps you could help me understand something else... as a Dynatrace representative, do you expect your customers to be using this `xray-lambda` propagator or even care about the active tracing spans generated by x-ray? I was given the impression that Dynatrace customers generally don't use OpenTelemetry instrumentation for Lambda but rather use Dynatrace internal instrumentation. Please help me understand specifically why this is important to you.

Reply:
> Specifically those span links should be on a message span, not the event span. 

The AWS Lambda invocation is IIUC a message "deliver" operation and linking the message senders to the Lambda invocation is sensible and implemented in at least some instrumentations.

>  That same document also specifies that link propagation should be done specifically using the x-ray propagator rather than the global propagator because x-ray is the only option AWS supports for more complicated interaction patterns

Right, but first it is "SHOULD" and for example Dynatrace requires the use of the global propagator (or specific other propagator) as the context information is transmitted using message attributes instead (additional tracestate attributes not transferred by the X-Ray propagator are required).

Second, why is it acceptable that that messaging propagator is configurable via an instrumentation propagator but for the event propagator it is not?

Reply:
If Dynatrace is requiring a specific other propagator, then I assume the `xray-lambda` propagator is likely not on the list or not first in priority.

> Second, why is it acceptable that that messaging propagator is configurable via an instrumentation propagator but for the event propagator it is not?

I'm fine if we decide to remove that requirement and instead just expect `x-ray` to be listed in the propagators configuration. I think it needs to be one or the other though... `x-ray` shouldn't be excluded from the propagation options.

Reply:
> Let's say I have a Lambda function that is timer-triggered. Every 10 minutes it polls all messages from some queue (maybe SQS, maybe a custom REST-based API). When the function is triggered, I want to have the X-Ray span as parent of the incoming FaaS span, so I set the global propagator to xray-lambda. So far so good. Now I (or the AWS SDK or custom REST messaging instrumentation) will call GetGlobalPropagator().Extract(message.headers) to extract a link to put on the message. Because the global propagator is configured to xray-lambda, this succeeds for every message. Every message will be linked to the X-Ray span for the Lambda handler (oops). If the messages actually had some other messaging headers that contained a span context, it will be lost.

this does seem like a potential problem

@Oberon00 would this behavior address the issue?

> [There is a separate effort to allow span contexts from additional propagators to be added as span links instead of just overwriting them.](https://github.com/open-telemetry/semantic-conventions/pull/354#issue-1919560575)

Reply:
@tylerbenson I had a typo here. I meant to write:

> Second, why is it acceptable that that messaging propagator is configurable via an instrumentation ~~propagator~~ <ins>option</ins> but for the event propagator it is not?

Reply:
@trask In this particular case, the link would normally have quite a few attributes (like message ID). Would these then be duplicated as well? It certainly seems doable although a bit weird and complex.

Reply:
I think the problem that this PR tries to solve (for an AWS-specific systems) is what we in the messaging SIG called ["context layers"](https://github.com/open-telemetry/oteps/blob/main/text/trace/0205-messaging-semantic-conventions-context-propagation.md#transport-context-propagation).

There's a _message context layer_ which is used to correlate producers with consumer, the _message context_ is usually attached to the message. And there's a _transport context layer_, which correlates producers and consumers with an intermediary. In the case of AWS, the _transport context_ is what's in the X-Ray environment variable.

If I understand correctly, this PR aims at covering both "context layers" with global propagators. @Oberon00 on the other hand proposes to cover the two "context layers" with two different propagator settings (global propagators for the _transport context layer_, instrumentation specific propagators for the _message context layer_).

While we identified the problem in the messaging SIG, we didn't come up with a solution (yet). We came up with conventions around span relationships on the _message context layer_, but we didn't touch the _transport context layer_, and we also didn't touch the relationship between context layers and global propagators.

Reply:
@Oberon00 I've added a clause stating the `xray-lambda` propagator should ignore if there's an active span in the context.  This doesn't solve the issue around extracting for span links as that might be done before a span is in context, but I think the other proposal I mentioned previously can help there.

Reply:
Not that I expect anyone to listen to me, but I'll state one more time for the record:  the instrumentation needs to provide control over how the carrier used to extract the parent span context is constructed, not dictate which propagator is used to extract that context.  @Oberon00 is absolutely correct that this is an instrumentation context carrier construction issue and cannot be solved cleanly by global propagator configuration.

Reply:
I agree that something with the propagator API needs to change, but I think the remaining concerns here should be resolved as part of a broader effort for how to enable context propagation generically across environment variables (as this problem isn't unique to the X-Ray environment variable propagation):
https://github.com/open-telemetry/opentelemetry-specification/issues/740
https://github.com/open-telemetry/oteps/pull/223

I'm going to resolve this conversation so we can proceed on unifying the implementations. If you feel something is being glossed over, I suggest starting in a new conversation on this PR as this chain is already too large.